[#11] Add OpenGraph Protocol Meta Tags

prikhi opened the pull request:
Add OpenGraph Protocol meta tags to the generated HTML. These describe
the site, streams, and topics to website's that provide rich media for
linked pages(e.g., Facebook, LinkedIn).

Streams simply display the Stream Name & Description.

Topics show the Topic Name, the Stream name, the created & modified
times, a description using the first 300 characters of the first
message, and the originals poster's avatar as an image.

A new `baseUrl` configuration value has been added, which should contain
the protocol & domain of the archive(e.g., `https://funprog.srid.ca`).
This is used to generate the `url` OGP tags, which require a full URI
instead of just a path.

Refs #11

prikhi committed:
[#11] Add OpenGraph Protocol Meta Tags

Add OpenGraph Protocol meta tags to the generated HTML. These describe
the site, streams, and topics to website's that provide rich media for
linked pages(e.g., Facebook, LinkedIn).

Streams simply display the Stream Name & Description.

Topics show the Topic Name, the Stream name, the created & modified
times, a description using the first 300 characters of the first
message, and the originals poster's avatar as an image.

A new `baseUrl` configuration value has been added, which should contain
the protocol & domain of the archive(e.g., `https://funprog.srid.ca`).
This is used to generate the `url` OGP tags, which require a full URI
instead of just a path.

Refs #11

prikhi commented:
Also my nix is broken atm so I wasn't able to test builds... I'll see how CI does :(

prikhi commented:
Alright, builds are fixed now. To finish this, I need an instance-agnostic way of getting the URL the archive will be hosted at since the `url` and `image` properties should be the full URL w/ protocol, domain, and path.

I might also clean up the generation of properties dependent on `Maybe a` types.

srid commented:
@prikhi You can get the base URL from the `cfg` value, which is passed as an argument to `generateSite` function. Just pass it down to `renderPage` and from there use `Config.baseUrl cfg`. 

The `Config` type itself is [generated by Dhall.TH](https://funprog.srid.ca/dhall/automatic-haskell-types.html), and you can check the available fields in `./config/Type.dhall`.

> Also my nix is broken atm so I wasn't able to test builds

Let me know (here or in Zulip) if there is any problem with getting the app running locally!

srid commented on src/Main.hs:
```diff
@@ -205,6 +206,45 @@ renderPage server page = with html_ [lang_ "en"] $ do
+                    [ property_ "og:description"
+                    , content_ $ T.take 300 $ _messageContent msg
+                    ]
+                  -- TODO: Is _messageAvatarUrl the full URL or just the path?
```
It's the full URL. 

(Perhaps I should change its type to `URI` to prevent this confusion in future).

srid commented on src/Main.hs:
```diff
@@ -205,6 +206,45 @@ renderPage server page = with html_ [lang_ "en"] $ do
+                  maybe (return ()) (\avatarUrl ->
+                    meta_ [property_ "og:image", content_ avatarUrl]
+                    ) $ _messageAvatarUrl msg
+              ) $ listToMaybe $ _topicMessages t
```
It's arguably better in terms of readability to replace the `maybe ...` here with a simple `case`, viz.:

```haskell
case listToMaybe (_topicMessages t) of 
  Nothing -> pure ()
  Just msg -> do 
    ...
```

Alternatively, since the app uses relude as its prelude, you can use [`whenJust`](https://hackage.haskell.org/package/relude-0.6.0.0/docs/Relude-Monad-Maybe.html#v:whenJust):

```haskell
whenJustM (listToMaybe $ _topicMessages t) $ \msg -> do
  ...
```

srid commented on src/Main.hs:
```diff
@@ -205,6 +206,45 @@ renderPage server page = with html_ [lang_ "en"] $ do
+                    ) $ _messageAvatarUrl msg
+              ) $ listToMaybe $ _topicMessages t
+    ogpTimeFormat =
+      T.pack
```
`T.pack` can be replaced with `toText` (from relude).

srid commented on src/Main.hs:
```diff
@@ -205,6 +206,45 @@ renderPage server page = with html_ [lang_ "en"] $ do
+            meta_ [property_ "og:title", content_ $ _topicName t]
+            meta_ [property_ "og:type", content_ "article"]
+            meta_ [property_ "og:article:section", content_ $ _streamName s]
+            -- TODO: Add a maybeProperty function?
```
I use `mapM_` to deal with maybe properties. See [example](https://github.com/srid/website/blob/a0e8db4/src/Website/View/OpenGraph.hs#L28).

prikhi commented:
> You can get the base URL from the cfg value, which is passed as an argument to generateSite function. Just pass it down to renderPage and from there use Config.baseUrl cfg.
> 
> The Config type itself is generated by Dhall.TH, and you can check the available fields in ./config/Type.dhall.

That's the Zulip domain right, I was thinking that the `url` property should point to the archive's domain instead(e.g., `funprog.srid.ca`). Is that available somewhere or should I add another config field for that?

prikhi commented:
Also, I'm not sure if it matters that there's HTML in the `description` tag.

srid commented:
> That's the Zulip domain right, I was thinking that the url property should point to the archive's domain instead(e.g., funprog.srid.ca). Is that available somewhere or should I add another config field for that?

srid commented:
> That's the Zulip domain right, I was thinking that the url property should point to the archive's domain instead(e.g., funprog.srid.ca). Is that available somewhere or should I add another config field for that?

Oops, my bad. Yea, you can just add another config field for that.

srid commented:
> Also, I'm not sure if it matters that there's HTML in the description tag.

Does it? I see that you are using `_streamDescription s`, which should just be a text.

prikhi commented:
The stream `description`s should be fine, but some of the topic descriptions have HTML like `<code>` elements.

Do you have a preference for naming of the archive domain field?

srid commented:
About the `<code>` elements, you can leave them as is for now. The zulip API actually does return the plain text version (Markdown) as well, I just haven't added that field to Zulip/Client.hs; I can do that after merging this PR (or you can poke around).

> Do you have a preference for naming of the archive domain field?

`baseUrl`.

prikhi commented:
* Added & used the `baseUrl` config option.
* Removed the TODOs & WIP from the commit & mentioned the new config option.

Let me know if there's anything else I should address.

prikhi commented:
Little more cleanup:

* fixed indents to 2 spaces
* remove the page parameter to the `ogpMetaTags` function since we can just use the `page` argument of the `renderPage` function.

srid commented:
@prikhi Thanks, and deployed! I changed the OGP title to match that of the page.
